pkg/puller/watcher.go

```diff
@@ -0,0 +1,184 @@
+package puller
+
+import (
+	"encoding/json"
+	"github.com/fsnotify/fsnotify"
+	"github.com/google/go-cmp/cmp"
+	"github.com/kubeflow/kfserving/pkg/apis/serving/v1beta1"
+	"github.com/kubeflow/kfserving/pkg/constants"
+	"github.com/kubeflow/kfserving/pkg/modelconfig"
+	"golang.org/x/sync/syncmap"
+	"io/ioutil"
+	"log"
+	"path/filepath"
+	"time"
+)
+
+var w *Watcher
+
+type Watcher struct {
+	configDir    string
+	ModelTracker *syncmap.Map
+	numWorkers   int
+}
+
+func init() {
+	w = NewWatcher()
+}
+
+func NewWatcher() *Watcher {
+	w := new(Watcher)
+	m := new(syncmap.Map)
+	w.ModelTracker = m
+	return w
+}
+
+type LoadState string
+
+const (
+	// State Related
+	ShouldLoad   LoadState = "Load"
+	ShouldUnload LoadState = "Unload"
+)
+
+type EventWrapper struct {
+	ModelSpec      *v1beta1.ModelSpec
+	LoadState      LoadState
+	ShouldDownload bool
+}
+
+type ModelWrapper struct {
+	ModelSpec *v1beta1.ModelSpec
+	Time      time.Time
+	Stale     bool
+	Success   bool
+}
+
+func WatchConfig(configDir string, numWorkers int) {
+	log.Println("Entering watch")
+	w.configDir = configDir
+	w.numWorkers = numWorkers
+	w.WatchConfig()
+}
+
+func (w *Watcher) WatchConfig() {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer watcher.Close()
+
+	done := make(chan bool)
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				isCreate := event.Op&fsnotify.Create != 0
+				eventPath := filepath.Clean(event.Name)
+				isDataDir := filepath.Base(eventPath) == "..data"
+				// TODO: Should we use atomic integer or timestamp??
+				timeNow := time.Now()
+				if isDataDir && isCreate {
+					symlink, _ := filepath.EvalSymlinks(eventPath)
+					file, err := ioutil.ReadFile(filepath.Join(symlink, constants.ModelConfigFileName))
+					modelConfigs := make(modelconfig.ModelConfigs, 0)
+					if err != nil {
+						log.Println("Error in reading file", err)
+					}
+					err = json.Unmarshal([]byte(file), &modelConfigs)
+					if err != nil {
+						log.Println("unable to marshall for", event, "with error", err)
+					}
+					for _, modelConfig := range modelConfigs {
```

Comment:
I think we need another loop that loops over the all the keys in ModelTracker and check it if is delete from the modelConfigs. It it is deleted, then remove from the ModelTracker.